src/transforms/DummyMainFunction.cpp

```diff
@@ -0,0 +1,262 @@
+                Constant* res = m_ndfn [type];
+                if (!res) {
+                    res = &makeNewNondetFn (M, *type, m_ndfn.size (), "verifier.nondet.");
+                    m_ndfn[type] = res;
+                }
+                return res;
+            }
+
+            std::map<std::string, std::vector<std::string>> getFnMap(Module& m){
+                std::map<std::string, std::vector<std::string>> fn_map;
+
+                for(auto &F: m){
+                    std::string demangled = demangle(F.getName());
+                    demangled = demangled.substr(0,demangled.find_first_of("("));
+                    
+                    if(fn_map.find(demangled) == fn_map.end()){
+                        std::vector<std::string> names;
+                        fn_map[demangled] = names;
+                    }
+
+                    fn_map[demangled].push_back(F.getName());
+                }
+
+                return fn_map;
+            }
+
+            void printFnMapInfo(std::map<std::string, std::vector<std::string>> fn_map){
+                errs()<<"|\tPrinting Function Name Map\t|\n";
+                errs()<<"_______________________________________________\n";
+                for(auto iter = fn_map.begin(),e = fn_map.end(); iter != e; iter++){
+                    std::string fn = iter->first;
+                    std::vector< std::string > vec = iter-> second;
+
+                    errs()<<fn<<":\n";
+                    for(auto nm: vec){
+                        errs()<<"\t"<<nm<<"\n";
+                    }
+                    errs()<<"_______________________________________________\n";
+                }
+            }
+
+            void makePrintf(Module& m, CallInst* ci, IRBuilder<> builder){
+
+                Type* charType = Type::getInt8PtrTy(m.getContext());
+                FunctionType *printf_type = FunctionType::get(charType,true);
+                auto *res = cast<Function>(m.getOrInsertFunction("printf", printf_type).getCallee());
+
+                assert(res && "printf not found in module");
+
+                Type* type = res->getReturnType();
+
+                // Testing with fixed type
+                std::string type_fmt = "%d";
+
+
+                SmallVector<Value*, 16> Args;
+
+                std::string fmt = "Creating print call for " + type_fmt;
+                Value* global_fmt = builder.CreateGlobalStringPtr(fmt.c_str());
+
+                Args.push_back(global_fmt);
+                Args.push_back((Value *) ci);
+
+                CallInst* printFunc = builder.CreateCall(res, Args);
+
+                errs() <<"DummyMainFunction: created print call "<< *printFunc << "\n";
+
+            }
+
+            public:
+
+            static char ID;
+
+            DummyMainFunction () : ModulePass (ID) {
+                // Initialize sea-dsa pass
+                llvm::PassRegistry &Registry = *llvm::PassRegistry::getPassRegistry();
```

Comment:
don't need this code since you are not trying to update the seadsa callgraph

Reply:
I had to create the printf calls, because otherwise the values weren’t being used in the main function and were being discarded during the specialization phases.

Reply:
yes, but you explicitly need to update the call graph if you want to keep it updated. The callgraph is not updated when a new function is added into a module. Since this pass will be run at the beginning you don't really need to keep updated anything.

Reply:
Sorry I'm a bit lost, should I just remove the code highlighted in this comment? If we need to update the callgraph, how would we go about it? I checked the llvm reference and should I just call the callgraph constructor with the module?

Reply:
Just remove the code that initialize sea-dsa. It's doing nothing.